java/core/src/main/java/com/microsoft/bond/StructBondType.java

```diff
@@ -0,0 +1,1359 @@
+        TaggedDeserializationContext context = new TaggedDeserializationContext(reader);
+        return this.deserializeValue(context);
+    }
+
+    @Override
+    final TypeDef createSchemaTypeDef(HashMap<StructBondType<?>, StructDefOrdinalTuple> structDefMap) {
+        // first need to get the struct def info since need to reference it by the ordinal
+        StructDefOrdinalTuple structDefInfo = structDefMap.get(this);
+        if (structDefInfo == null) {
+            // struct def wasn't created yet, create one and associate with the current struct in the map;
+            int nextOrdinal = structDefMap.size();
+            StructDef structDef = new StructDef();
+            structDefInfo = new StructDefOrdinalTuple(structDef, nextOrdinal);
+
+            // the struct def instance that is associated in the map with the current struct type is not yet
+            // initialized, but any descendants that use this struct will reference the same struct def instance
+            structDefMap.put(this, structDefInfo);
+            this.initializeSchemaStructDef(structDef, structDefMap);
+        }
+
+        TypeDef typeDef = new TypeDef();
+        typeDef.id = this.getBondDataType();
+        typeDef.struct_def = (short) structDefInfo.ordinal;
+        return typeDef;
+    }
+
+    /**
+     * Codegen helper method that tries to read a field from payload and indicates whether there is a field to read.
+     *
+     * @param context contains the runtime context of the deserialization
+     * @return true if there are more fields to read
+     * @throws IOException if an I/O error occurred
+     */
+    protected static boolean readField(TaggedDeserializationContext context) throws IOException {
+        context.reader.readFieldBegin(context.readFieldResult);
+        int statusValue = context.readFieldResult.type.value;
+        return statusValue != BondDataType.BT_STOP.value && statusValue != BondDataType.BT_STOP_BASE.value;
+    }
+
+    /**
+     * Resolves generic type arguments and caches the result in the type cache.
+     *
+     * @param resolver             type resolver
+     * @param genericTypeArguments generic type arguments
+     * @param <TStruct>            the class of the struct value
+     * @return a cached uninitialized type descriptor
+     */
+    protected static <TStruct extends BondSerializable> StructBondType<TStruct> resolveUninitializedWithCaching(
+            StructBondTypeResolver<TStruct> resolver, BondType<?>... genericTypeArguments) {
+        BondType<?>[] cachedGenericTypeArguments = new BondType[genericTypeArguments.length];
+        for (int i = 0; i < genericTypeArguments.length; ++i) {
+            cachedGenericTypeArguments[i] = typeCache.get(genericTypeArguments[i]);
+        }
+        StructBondType<TStruct> type = resolver.resolveUninitialized(cachedGenericTypeArguments);
+        return (StructBondType<TStruct>) BondType.typeCache.get(type);
+    }
+
+    private void initializeSchemaStructDef(
```

Comment:
There's a public method "createSchemaDef" in StructBondType. It's not called yet, and we'll take a closer look when designing untagged protocols.

---
In reply to: [123399536](https://github.com/Microsoft/bond/pull/502#discussion_r123399536) [](ancestors = 123399536)